presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java

```diff
@@ -496,6 +506,149 @@ public CounterStat getHighMemorySplitSource()
+                }
+                boolean pruned = false;
+                if (partitionStatistics.containsKey(entry.getKey())) {
+                    Map<String, HiveColumnStatistics> columnStatsMap = partitionStatistics.get(entry.getKey()).getColumnStatistics();
+                    for (Map.Entry<String, HiveColumnHandle> predicateColumnEntry : predicateColumns.entrySet()) {
+                        if (columnStatsMap.containsKey(predicateColumnEntry.getKey())) {
+                            Optional<ValueSet> columnsStatsValueSet = getColumnStatsValueSet(columnStatsMap.get(predicateColumnEntry.getKey()), predicateColumnEntry.getValue().getHiveType());
+                            if (columnsStatsValueSet.isPresent()) {
+                                ValueSet columnPredicateValueSet = domains.get().get(new Subfield(predicateColumnEntry.getKey())).getValues();
+                                if (!columnPredicateValueSet.overlaps(columnsStatsValueSet.get())) {
+                                    pruned = true;
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                }
+
+                if (pruned) {
+                    prunedPartitionNamesBuilder.add(entry.getKey());
+                }
+                else {
+                    partitionBuilder.put(entry.getKey(), entry.getValue().get());
+                }
+            }
+            return new PartitionBatchInfo(partitionBuilder.build(), prunedPartitionNamesBuilder.build());
+        }
+
+        Map<String, Optional<Partition>> batch = metastore.getPartitionsByNames(
+                tableName.getSchemaName(),
+                tableName.getTableName(),
+                Lists.transform(partitionBatch, HivePartition::getPartitionId));
+        ImmutableMap.Builder<String, Partition> partitionBuilder = ImmutableMap.builder();
+        for (Map.Entry<String, Optional<Partition>> entry : batch.entrySet()) {
+            if (!entry.getValue().isPresent()) {
+                throw new PrestoException(HIVE_PARTITION_DROPPED_DURING_QUERY, "Partition no longer exists: " + entry.getKey());
+            }
+            partitionBuilder.put(entry.getKey(), entry.getValue().get());
+        }
+        return new PartitionBatchInfo(partitionBuilder.build());
+    }
+
+    private static Optional<ValueSet> getColumnStatsValueSet(HiveColumnStatistics statistics, HiveType hiveType)
+    {
+        // TODO: Support more types
+        if (hiveType.equals(HIVE_BYTE) || hiveType.equals(HIVE_SHORT) || hiveType.equals(HIVE_INT) || hiveType.equals(HIVE_LONG)) {
+            return getIntegerColumnStatsValueSet(statistics, hiveType);
+        }
+        else {
+            return Optional.empty();
+        }
+    }
+
+    private static Optional<ValueSet> getIntegerColumnStatsValueSet(HiveColumnStatistics statistics, HiveType hiveType)
```

Comment:
Let's reuse `MetastoreHiveStatisticsProvider`'s `isRangeSupported` and `createRange` helpers to make it very generic. We can also return `DoubleRange` instead of `ValueSet` of simplicity. Likely we can ignore or add TODO for non-numeric types.

Reply:
Oh wow very cool. Not aware of it. Thanks for pointing this out

Reply:
As discussed, we likely will not use `isRangeSupported` because it will cast data into double, which might lead to accuracy lost if we have large Long statistics.